.github/workflows/test.yml

```diff
@@ -0,0 +1,90 @@
+name: Test
+
+on: push
+
+env:
+  DOCKER_HUB_USERNAME: ${{ github.repository_owner }}
+  DOCKER_HUB_TOKEN: ${{ secrets.DOCKER_HUB_TOKEN }}
+  COMPOSE_FILE: docker-compose.test.yml
+  DOCKER_IMAGE: ${{ github.repository }}
+
+jobs:
+  build:
+    name: Build docker image
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: ./.github/actions/branch-tag
+
+      - name: Docker login
+        run: echo "$DOCKER_HUB_TOKEN" | docker login --username=$DOCKER_HUB_USERNAME --password-stdin
+
+      - name: Docker pull
+        if: env.BRANCH_TAG != 'master' && env.BRANCH_TAG != 'development'
+        run: docker-compose pull test
+
+      - name: Docker build
+        run: |
+          bin/docker-prepare
+          docker-compose build
+          docker-compose push test
+
+  unit_test:
+    name: Unit test
+    runs-on: ubuntu-latest
+    needs: build
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: ./.github/actions/branch-tag
+
+      - name: Docker login
+        run: echo "$DOCKER_HUB_TOKEN" | docker login --username=$DOCKER_HUB_USERNAME --password-stdin
+
+      - name: Docker pull
+        run: docker-compose pull
+
+      - name: Linters
+        run: docker-compose run test yarn lint
+
+      - name: RSpec
+        run: docker-compose run test bundle exec rspec --exclude-pattern "spec/systems/**/*_spec.rb" --profile
+
+  system_test:
+    name: System test
+    runs-on: ubuntu-latest
+    needs: build
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: ./.github/actions/branch-tag
+
+      - name: Docker login
+        run: echo "$DOCKER_HUB_TOKEN" | docker login --username=$DOCKER_HUB_USERNAME --password-stdin
+
+      - name: Docker pull
+        run: docker-compose pull
+
+      - name: RSpec
+        run: docker-compose run test bundle exec rspec spec/systems --profile
+
+  deploy:
+    name: Invoke deployment
+    runs-on: ubuntu-latest
+    needs: [unit_test, system_test]
```

Comment:
Sorry but I am still not convinced n the current deployment config. I am not sure if we need `benc-uk/workflow-dispatch` especially since it's a fairly new action plugin:

![image](https://user-images.githubusercontent.com/696529/88256934-c5b8fa00-cce6-11ea-9e60-70d09a2ae09c.png)

There could be a check on the `github.event`:

```
staging-deploy:
  name: Stafing deployment
  runs-on: ubuntu-latest
  if: github.event.pull_request.merged && contains(github.ref, 'refs/heads/development')
```

But I still think it should be in another deployment workflow (not in the test workflow):

```
// in deployment.yml
on:
  push:
    branches:
      - master
      - develop

steps:
  - name: Staging deployment
    if: github.ref == 'refs/heads/develop'

   - name: Production deployment
    if: github.ref == 'refs/heads/master'
  
```

Reply:
I want the deployment to be run only after passed the tests, and I also wanna trigger the deployment manually, then I use `workflow_dispatch` event. I need `benc-uk/workflow-dispatch` to call an API that dispatch another workflow by using a `PERSONAL_TOKEN`.

If using the `workflow_dispatch` event, we will have this on the UI:
![Actions_·_hoangmirs_test-repo](https://user-images.githubusercontent.com/7344405/88265880-7085e400-ccf8-11ea-9541-95d4e7d43876.png)

And we can deploy manually any branches to the server

Reply:
I changed to use the `push` event instead of `workflow_dispatch`.
Currently, I still keep 2 files for staging and production as I use different Heroku servers.